src/scenic/core/specifiers.py

```diff
@@ -3,79 +3,119 @@
-		self.value = value
-
-		def enabled(thing, default):
-			if thing in attributes:
-				attributes.remove(thing)
-				return True
-			else:
-				return default
-		self.isAdditive = enabled('additive', False)
-		self.isDynamic = enabled('dynamic', False)
-		for attr in attributes:
-			raise RuntimeParseError(f'unknown property attribute "{attr}"')
-
-	@staticmethod
-	def forValue(value):
-		if isinstance(value, PropertyDefault):
-			return value
-		else:
-			return PropertyDefault(set(), set(), lambda self: value)
-
-	def resolveFor(self, prop, overriddenDefs):
-		"""Create a Specifier for a property from this default and any superclass defaults."""
-		if self.isAdditive:
-			allReqs = self.requiredProperties
-			for other in overriddenDefs:
-				allReqs |= other.requiredProperties
-			def concatenator(context):
-				allVals = [self.value(context)]
-				for other in overriddenDefs:
-					allVals.append(other.value(context))
-				return tuple(allVals)
-			val = DelayedArgument(allReqs, concatenator, _internal=True)
-		else:
-			val = DelayedArgument(self.requiredProperties, self.value, _internal=True)
-		return Specifier(prop, val)
+    """A default value, possibly with dependencies."""
+    def __init__(self, requiredProperties, attributes, value):
+        self.requiredProperties = set(requiredProperties)
+        self.value = value
+
+        def enabled(thing, default):
+            if thing in attributes:
+                attributes.remove(thing)
+                return True
+            else:
+                return default
+
+        self.isAdditive = enabled('additive', False)
+        self.isDynamic = enabled('dynamic', False)
+        self.isFinal = enabled('final', False)
+        for attr in attributes:
+            raise RuntimeError(f'unknown property attribute "{attr}"')
+
+    @staticmethod
+    def forValue(value):
+        if isinstance(value, PropertyDefault):
+            return value
+        else:
+            return PropertyDefault(set(), set(), lambda self: value)
+
+    def resolveFor(self, prop, overriddenDefs):
+        """Create a Specifier for a property from this default and any superclass defaults."""
+        for other in overriddenDefs:
+            if other.isFinal:
+                msg = f'"{prop}" property cannot be overridden'
+                if prop == 'heading':
+                    msg += ' (perhaps this scenario requires the --2d flag?)'
+                raise InvalidScenarioError(msg)
+        if self.isAdditive:
+            allReqs = self.requiredProperties
+            for other in overriddenDefs:
+                allReqs |= other.requiredProperties
+            def concatenator(context):
+                allVals = [self.value(context)]
+                for other in overriddenDefs:
+                    allVals.append(other.value(context))
+                return tuple(allVals)
+            val = DelayedArgument(allReqs, concatenator, _internal=True) # TODO: @Matthew Change to dicts
```

Comment:
Not sure what this comment was referring to, but it looks like line 121 properly creates a dict and the tests for additive properties are passing, so I think we can remove it.